### PR TITLE
4765: Guest users are allowed to read data from a secondary datasource

### PR DESCRIPTION
**Pull request by tkmonson**
Added support for a second datasource that will point to the master database as well as any secondary replicas. This involved:

- Adding a second persistence unit in persistence.xml and changing the name of the original persistence unit.

- Adding a second datasource and second connection pool in glassfish-setup.sh.

- Writing chooseEM() in DatasetServiceBean to determine when to read from a primary or secondary.

- Many other additions and little tweaks to names of variables.

Currently, this is only being used in the event of a guest user reading a published dataset, but reading from a secondary datasource could easily be implemented for many other events.

Ideally, the second EntityManager should be created only if it is needed (i.e. within a conditional branch in chooseEM). However, I decided to stay consistent with the rest of the code, which configures persistence units in persistence.xml and injects EntityManagers via the @PersistenceContext annotation. It might be possible to instantiate EntityManagers dynamically using an EntityManagerFactory, but I'm not convinced that would be a better solution. More info on both of these approachs [here](https://stackoverflow.com/a/24572869).

connects to #4765

### CODE REVIEW

**Comment by oscardssmith** [comment]
What's the perf impact?

**Comment by djbrooke** [comment]
Hi @tkmonson - is this ready for code review?

**Comment by tkmonson** [comment]
@djbrooke Yes, it's ready for review, but @danmcp told me he would like to look at the changes before it is merged, so just keep that in mind.

@oscardssmith I'm running some timed tests right now. It should improve availability for reads of datasets. I'll keep you updated.

**Comment by danmcp** [comment]
@djbrooke @oscardssmith @tkmonson A slight clarification.  This change is only one of many to fully enable secondary reads for unauthenticated requests.  It sets up the basic framework to allow secondary requests but doesn't change all the places to take advantage.  The intent is to validate the methodology before continuing with changing all the callers which should all be done in this same PR.

**Comment by danmcp** [comment]
@tkmonson Please squash the commits.

File: conf/docker-aio/testscripts/install [hunk 1]
```diff
@@ -5,6 +5,7 @@ export FILES_DIR=/usr/local/glassfish4/glassfish/domains/domain1/files
 export DB_NAME=dvndb
 export DB_PORT=5432
 export DB_HOST=localhost
+export DB_SERVICE=dataverse-postgresql-service
```
Comment (danmcp) [inline conf/docker-aio/testscripts/install]: Does this relate at all to the kube code path?

File: conf/openshift/openshift.json [hunk 2]
```diff
@@ -162,7 +163,7 @@
         "name": "iqss-dataverse-solr"
       },
       "spec": {
-        "dockerImageRepository": "iqss/dataverse-solr"
+        "dockerImageRepository": "iqss/dataverse-solr:4762"
```
Comment (danmcp) [inline conf/openshift/openshift.json]: huh?

File: conf/openshift/openshift.json [hunk 3]
```diff
@@ -105,6 +105,7 @@
         "selector": {
           "name": "iqss-dataverse-solr"
         },
+        "clusterIP": "None",
```
Comment (danmcp) [inline conf/openshift/openshift.json]: Why?

File: conf/openshift/openshift.json [hunk 4]
```diff
@@ -249,7 +250,7 @@
             "containers": [
               {
                 "name": "dataverse-plus-glassfish",
-                "image": "iqss/dataverse-glassfish:latest",
+                "image": "tmonson/dataverse-glassfish:latest",
```
Comment (danmcp) [inline conf/openshift/openshift.json]: Can't commit this change.  Need to undo it for the PR.

File: conf/openshift/openshift.json [hunk 5]
```diff
@@ -526,15 +527,16 @@
       }
     },
     {
-      "kind": "DeploymentConfig",
-      "apiVersion": "v1",
+      "kind": "StatefulSet",
```
Comment (danmcp) [inline conf/openshift/openshift.json]: Why is this in your pull?

File: scripts/installer/glassfish-startup [hunk 6]
```diff
@@ -43,6 +43,7 @@ my @CONFIG_VARIABLES;
 	'MAIL_SERVER',
 
 	'POSTGRES_SERVER',
+        'POSTGRES_SERVICE',
```
Comment (danmcp) [inline scripts/installer/glassfish-startup]: formatting is off

File: scripts/installer/glassfish-setup.sh [hunk 7]
```diff
@@ -163,6 +170,12 @@ if [ -z "$DB_HOST" ]
   exit 1
 fi
 
+if [ -z "$DB_SERVICE" ]
```
Comment (danmcp) [inline scripts/installer/glassfish-setup.sh]: I don't think we want to add this requirement.  We should just not setup the extra datasource is DB_SERVICE isn't specified.

File: scripts/installer/install [hunk 8]
```diff
@@ -53,6 +53,7 @@ else
 	'MAIL_SERVER',
 
 	'POSTGRES_SERVER',
+        'POSTGRES_SERVICE',
```
Comment (danmcp) [inline scripts/installer/install]: formatting is off

File: scripts/installer/postgres-setup [hunk 9]
```diff
@@ -31,13 +32,21 @@ my %CONFIG_DEFAULTS = (
...
     'POSTGRES_SERVER',   'Postgres Server Address',
+    'POSTGRES_SERVICE',  'Postgres Service Address',
     'POSTGRES_PORT',     'Postgres Server Port',
     'POSTGRES_ADMIN_PASSWORD', 'Postgres ADMIN password',
     'POSTGRES_DATABASE', 'Name of the Postgres Database',
     'POSTGRES_USER',     'Name of the Postgres User',
     'POSTGRES_PASSWORD', 'Postgres user password',
 );
 
+$CONFIG_DEFAULTS{'POSTGRES_PORT'} = $ENV{'DB_PORT'};
```
Comment (danmcp) [inline scripts/installer/postgres-setup]: Why was this necessary for all the existing values?

File: scripts/issues/2595/monitor.py [hunk 10]
```diff
@@ -5,7 +5,7 @@
 import csv
 import os
 from shutil import move
-base_url = 'http://localhost:4848/monitoring/domain/server/resources/dvnDbPool'
+base_url = 'http://localhost:4848/monitoring/domain/server/resources/dvnDbPool1'
```
Comment (danmcp) [inline scripts/issues/2595/monitor.py]: Did you rename the existing pool?  We probably don't want to do that.

File: src/main/java/edu/harvard/iq/dataverse/DatasetServiceBean.java [hunk 11]
```diff
@@ -93,14 +94,27 @@
...
+    public EntityManager chooseEM(User user) {
+        if(user.isAuthenticated())  {
+            return em;
+        }
+        return em2;
+    }
 
     public Dataset find(Object pk) {
-        return em.find(Dataset.class, pk);
+        return chooseEM(session.getUser()).find(Dataset.class, pk);
```
Comment (danmcp) [inline src/main/java/edu/harvard/iq/dataverse/DatasetServiceBean.java]: getEntityManager is probably a better name.

File: src/main/java/edu/harvard/iq/dataverse/DatasetServiceBean.java [hunk 12]
```diff
@@ -93,14 +94,27 @@
...
+    DataverseSession session;
 
     private static final SimpleDateFormat logFormatter = new SimpleDateFormat("yyyy-MM-dd'T'HH-mm-ss");
     
-    @PersistenceContext(unitName = "VDCNet-ejbPU")
+    @PersistenceContext(unitName = "VDCNet-ejbPU-1")
     protected EntityManager em;
+    
+    @PersistenceContext(unitName = "VDCNet-ejbPU-2")
+    protected EntityManager em2;
```
Comment (danmcp) [inline src/main/java/edu/harvard/iq/dataverse/DatasetServiceBean.java]: Can we name this better?  Like maybe allowsSecondaryEM

File: src/main/resources/META-INF/persistence.xml [hunk 13]
```diff
@@ -1,9 +1,9 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <persistence version="1.0" xmlns="http://java.sun.com/xml/ns/persistence" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://java.sun.com/xml/ns/persistence http://java.sun.com/xml/ns/persistence/persistence_1_0.xsd">
-    <persistence-unit name="VDCNet-ejbPU" transaction-type="JTA">
+    <persistence-unit name="VDCNet-ejbPU-1" transaction-type="JTA">
```
Comment (danmcp) [inline src/main/resources/META-INF/persistence.xml]: Can we pick a better name for this rather than just -1

File: src/main/resources/META-INF/persistence.xml [hunk 14]
```diff
@@ -1,9 +1,9 @@
...
         <!-- provider>oracle.toplink.essentials.ejb.cmp3.EntityManagerFactoryProvider</provider-->
         <provider>org.eclipse.persistence.jpa.PersistenceProvider</provider>
-        <jta-data-source>jdbc/VDCNetDS</jta-data-source>
+        <jta-data-source>jdbc/VDCNetDS1</jta-data-source>
```
Comment (danmcp) [inline src/main/resources/META-INF/persistence.xml]: Same comment with names.

File: src/main/resources/META-INF/persistence.xml [hunk 15]
```diff
@@ -20,6 +20,26 @@
             <!--<property name="eclipselink.logging.parameters" value="true"/>-->
         </properties>
     </persistence-unit>
+    <persistence-unit name="VDCNet-ejbPU-2" transaction-type="JTA">
```
Comment (danmcp) [inline src/main/resources/META-INF/persistence.xml]: Same comment with names.

File: src/main/resources/META-INF/persistence.xml [hunk 16]
```diff
@@ -20,6 +20,26 @@
...
+        <!-- provider>oracle.toplink.essentials.ejb.cmp3.EntityManagerFactoryProvider</provider-->
+        <provider>org.eclipse.persistence.jpa.PersistenceProvider</provider>
+        <jta-data-source>jdbc/VDCNetDS2</jta-data-source>
```
Comment (danmcp) [inline src/main/resources/META-INF/persistence.xml]: Same comment with names.

**Comment by oscardssmith** [comment]
Have you guys done any performance testing yet or is it too early?

File: conf/openshift/openshift.json [hunk 17]
```diff
@@ -426,7 +547,7 @@
             "containers": [
               {
                 "name": "iqss-dataverse-solr",
-                "image": "iqss-dataverse-solr:latest",
```
Comment (danmcp) [inline conf/openshift/openshift.json]: You still seem to be pulling in changes from other commits.

File: conf/docker-aio/testscripts/install (hunk 1 above)
Comment (tkmonson) [reply conf/docker-aio/testscripts/install]: Removed.

File: conf/openshift/openshift.json (hunk 2 above)
Comment (tkmonson) [reply conf/openshift/openshift.json]: This is an unintentional change that came from a commit I cherry-picked. Removed.

File: conf/openshift/openshift.json (hunk 3 above)
Comment (tkmonson) [reply conf/openshift/openshift.json]: This is an unintentional change that came from a commit I cherry-picked. Removed.

File: conf/openshift/openshift.json (hunk 4 above)
Comment (tkmonson) [reply conf/openshift/openshift.json]: Fixed.

File: conf/openshift/openshift.json (hunk 5 above)
Comment (tkmonson) [reply conf/openshift/openshift.json]: This is an unintentional change that came from a commit I cherry-picked. Removed.

File: scripts/installer/glassfish-startup (hunk 6 above)
Comment (tkmonson) [reply scripts/installer/glassfish-startup]: Tabs vs spaces problem. Fixed.

File: scripts/installer/glassfish-setup.sh (hunk 7 above)
Comment (tkmonson) [reply scripts/installer/glassfish-setup.sh]: Removed.

File: scripts/installer/install (hunk 8 above)
Comment (tkmonson) [reply scripts/installer/install]: Tabs vs spaces problem. Fixed.

File: scripts/installer/postgres-setup (hunk 9 above)
Comment (tkmonson) [reply scripts/installer/postgres-setup]: Most of this code was duplicated a few lines lower, so I deleted duplicate code in this block.

File: scripts/issues/2595/monitor.py (hunk 10 above)
Comment (tkmonson) [reply scripts/issues/2595/monitor.py]: Fixed.

File: src/main/java/edu/harvard/iq/dataverse/DatasetServiceBean.java (hunk 11 above)
Comment (tkmonson) [reply src/main/java/edu/harvard/iq/dataverse/DatasetServiceBean.java]: Fixed.

File: src/main/java/edu/harvard/iq/dataverse/DatasetServiceBean.java (hunk 12 above)
Comment (tkmonson) [reply src/main/java/edu/harvard/iq/dataverse/DatasetServiceBean.java]: Renamed to slavePU.

File: src/main/resources/META-INF/persistence.xml (hunk 13 above)
Comment (tkmonson) [reply src/main/resources/META-INF/persistence.xml]: Renamed to masterPU.

File: src/main/resources/META-INF/persistence.xml (hunk 14 above)
Comment (tkmonson) [reply src/main/resources/META-INF/persistence.xml]: Renamed to masterDS.

File: src/main/resources/META-INF/persistence.xml (hunk 15 above)
Comment (tkmonson) [reply src/main/resources/META-INF/persistence.xml]: Renamed to slaveDS.

File: src/main/resources/META-INF/persistence.xml (hunk 16 above)
Comment (tkmonson) [reply src/main/resources/META-INF/persistence.xml]: Renamed to slavePU / slavePS.

**Comment by tkmonson** [comment]
Ok, commits have been squashed, unnecessary files have been removed, and I believe I have addressed all of the comments from @danmcp.

**Comment by tkmonson** [comment]
@danmcp and I have agreed that this PR should not be merged until more requests for data are redirected to the replica database. Currently there is one case where this happens, but I've found many more potential cases. I'll be pushing those changes over the next few days.

**Comment by tkmonson** [comment]
Ok, I pushed my most recent changes. The changes mostly involve injecting a new file called EntityManagerBean into each file that uses an EntityManager instead of creating a new EntityManager in each file with the PersistenceContext annotation. Where the EntityManager.find() method is called, the EntityManager is selected using getEntityManager() in EntityManagerBean. This chooses an EntityManager based on whether the session user is authenticated or not (auth = master database, unauth = either master or replica database).

However, lines 522-523 in DatasetServiceBean.java still use an EntityManager rather than reference EntityManagerBean. When I try to change this, two errors are thrown in two test files. I haven't been able to figured out what the problem is. [This branch](https://github.com/tkmonson/dataverse/tree/4765_scaled_database_test) contains the changes that throw the errors.

That said, if that problem is solved or deemed insignificant, then this pull request is ready for review.

File: conf/docker-aio/dv/install/default.config [hunk 18]
```diff
@@ -4,6 +4,7 @@ ADMIN_EMAIL
 MAIL_SERVER	mail.hmdc.harvard.edu
 POSTGRES_ADMIN_PASSWORD	secret
 POSTGRES_SERVER	db
+POSTGRES_SERVICE	dataverse-postgresql-service
```
[the remark this answers is not in the document]
Comment (tkmonson) [reply conf/docker-aio/dv/install/default.config]: Removed.

File: doc/sphinx-guides/source/_static/util/default.config [hunk 19]
```diff
@@ -4,6 +4,7 @@ ADMIN_EMAIL
 MAIL_SERVER	mail.hmdc.harvard.edu
 POSTGRES_ADMIN_PASSWORD	secret
 POSTGRES_SERVER	127.0.0.1
+POSTGRES_SERVICE	dataverse-postgresql-service
```
[the remark this answers is not in the document]
Comment (tkmonson) [reply doc/sphinx-guides/source/_static/util/default.config]: Removed.

File: scripts/installer/glassfish-startup [hunk 20]
```diff
@@ -69,6 +70,7 @@ my %CONFIG_DEFAULTS = (
 
     'POSTGRES_ADMIN_PASSWORD', 'secret',
     'POSTGRES_SERVER',   '127.0.0.1',
+    'POSTGRES_SERVICE',  'dataverse-postgresql-service',
```
[the remark this answers is not in the document]
Comment (tkmonson) [reply scripts/installer/glassfish-startup]: Changed to 127.0.0.1

File: scripts/installer/postgres-setup [hunk 21]
```diff
@@ -15,13 +15,14 @@ my $WORKING_OS = "";
 my $postgresonly = 0;
 
     @CONFIG_VARIABLES =
-      ( 'POSTGRES_SERVER', 'POSTGRES_PORT', 'POSTGRES_DATABASE', 'POSTGRES_USER', 'POSTGRES_PASSWORD', 'POSTGRES_ADMIN_PASSWORD' );
+      ( 'POSTGRES_SERVER', 'POSTGRES_SERVICE', 'POSTGRES_PORT', 'POSTGRES_DATABASE', 'POSTGRES_USER', 'POSTGRES_PASSWORD', 'POSTGRES_ADMIN_PASSWORD' );
```
[the remark this answers is not in the document]
Comment (tkmonson) [reply scripts/installer/postgres-setup]: Removed.

File: scripts/installer/install [hunk 22]
```diff
@@ -78,13 +79,14 @@ my %CONFIG_DEFAULTS = (
     'MAIL_SERVER',         'mail.hmdc.harvard.edu',
 
     'POSTGRES_ADMIN_PASSWORD', 'secret',
-    'POSTGRES_SERVER',   '127.0.0.1',
-    'POSTGRES_PORT',     5432,
-    'POSTGRES_DATABASE', 'dvndb',
-    'POSTGRES_USER',     'dvnapp',
-    'POSTGRES_PASSWORD', 'secret',
+    'POSTGRES_SERVER',         '127.0.0.1',
+    'POSTGRES_SERVICE',        'dataverse-postgresql-service',
```
[the remark this answers is not in the document]
Comment (tkmonson) [reply scripts/installer/install]: Changed to 127.0.0.1

File: scripts/installer/glassfish-startup [hunk 23]
```diff
@@ -89,6 +91,7 @@ my %CONFIG_DEFAULTS = (
 
     $CONFIG_DEFAULTS{'POSTGRES_PORT'} = $ENV{'DB_PORT'};
     $CONFIG_DEFAULTS{'POSTGRES_SERVER'} = $ENV{'DB_HOST'};
+    $CONFIG_DEFAULTS{'POSTGRES_SERVICE'} = $ENV{'DB_SERVICE'};
```
[the remark this answers is not in the document]
Comment (tkmonson) [reply scripts/installer/glassfish-startup]: Added.

File: scripts/installer/postgres-setup [hunk 24]
```diff
@@ -57,20 +57,23 @@ $pod_name = $ENV{'MY_POD_NAME'};
 if ($pod_name eq "start-glassfish")
 {
     print "Init container starting \n";
+
     $CONFIG_DEFAULTS{'POSTGRES_SERVER'} = $ENV{"POSTGRES_SERVER"} . "." .  $ENV{"POSTGRES_SERVICE_HOST"};
     $CONFIG_DEFAULTS{'POSTGRES_DATABASE'} = $ENV{"POSTGRES_DATABASE"};
     $CONFIG_DEFAULTS{'POSTGRES_USER'} = $ENV{"POSTGRES_USER"};
-    $CONFIG_DEFAULTS{'POSTGRES_ADMIN_PASSWORD'} = $ENV{"POSTGRES_ADMIN_PASSWORD"};
     $CONFIG_DEFAULTS{'POSTGRES_PASSWORD'} = $ENV{"POSTGRES_PASSWORD"};
+
```
Comment (danmcp) [inline scripts/installer/postgres-setup]: I am not sure why you are making all the changes in this section.

File: scripts/database/homebrew/devinstall [hunk 25]
```diff
@@ -3,6 +3,7 @@ cd scripts/installer
 export DB_NAME=dataverse_db
 export DB_PORT=5432
 export DB_HOST=localhost
+export DB_SERVICE=dataverse-postgresql-service
```
Comment (danmcp) [inline scripts/database/homebrew/devinstall]: This one looks wrong still

File: scripts/deploy/phoenix.dataverse.org/install [hunk 26]
```diff
@@ -5,6 +5,7 @@ export FILES_DIR=/usr/local/glassfish4/glassfish/domains/domain1/files
 export DB_NAME=dvndb
 export DB_PORT=5432
 export DB_HOST=localhost
+export DB_SERVICE=dataverse-postgresql-service
```
Comment (danmcp) [inline scripts/deploy/phoenix.dataverse.org/install]: this one looks wrong still

File: scripts/installer/glassfish-startup [hunk 27]
```diff
@@ -89,6 +91,14 @@ my %CONFIG_DEFAULTS = (
 
     $CONFIG_DEFAULTS{'POSTGRES_PORT'} = $ENV{'DB_PORT'};
     $CONFIG_DEFAULTS{'POSTGRES_SERVER'} = $ENV{'DB_HOST'};
+
+    if (defined $ENV{'DB_SERVICE'}) {
```
Comment (danmcp) [inline scripts/installer/glassfish-startup]: You might want to also check that it's not an empty string

File: scripts/installer/glassfish-startup [hunk 28]
```diff
@@ -349,11 +359,12 @@ sub setup_glassfish {
...
     $ENV{'MEM_HEAP_SIZE'}    = $CONFIG_DEFAULTS{'DEF_MEM_SIZE'};
 
-    $ENV{'DB_PORT'} = $CONFIG_DEFAULTS{'POSTGRES_PORT'};
-    $ENV{'DB_HOST'} = $CONFIG_DEFAULTS{'POSTGRES_SERVER'};
-    $ENV{'DB_NAME'} = $CONFIG_DEFAULTS{'POSTGRES_DATABASE'};
-    $ENV{'DB_USER'} = $CONFIG_DEFAULTS{'POSTGRES_USER'};
-    $ENV{'DB_PASS'} = $CONFIG_DEFAULTS{'POSTGRES_PASSWORD'};
+    $ENV{'DB_PORT'}    = $CONFIG_DEFAULTS{'POSTGRES_PORT'};
+    $ENV{'DB_HOST'}    = $CONFIG_DEFAULTS{'POSTGRES_SERVER'};
+    $ENV{'DB_SERVICE'} = $CONFIG_DEFAULTS{'POSTGRES_SERVICE'};
```
Comment (danmcp) [inline scripts/installer/glassfish-startup]: Shouldn't this one also be conditional?

File: scripts/setup/asadmin-setup.sh [hunk 29]
```diff
@@ -30,6 +30,7 @@ DEFAULT_ASADMIN_OPTS=" "
 #
 DB_PORT=5432; export DB_PORT
 DB_HOST=localhost; export DB_HOST
+DB_SERVICE=dataverse-postgresql-service; export DB_SERVICE
```
Comment (danmcp) [inline scripts/setup/asadmin-setup.sh]: This one seems wrong

File: scripts/installer/postgres-setup (hunk 24 above)
Comment (tkmonson) [reply scripts/installer/postgres-setup]: This appears to be me being obsessive about code format. Not necessary, removed changes.

File: scripts/database/homebrew/devinstall (hunk 25 above)
Comment (tkmonson) [reply scripts/database/homebrew/devinstall]: Changed to localhost.

File: scripts/deploy/phoenix.dataverse.org/install (hunk 26 above)
Comment (tkmonson) [reply scripts/deploy/phoenix.dataverse.org/install]: Changed to localhost.

**Comment by pdurbin** [comment]
Closing since this pull request has been put on ice per https://github.com/IQSS/dataverse/issues/4765#issuecomment-421064799
